libdnf5/rpm/solv/goal_private.cpp

```diff
@@ -262,61 +268,55 @@ bool GoalPrivate::limit_installonly_packages(libdnf5::solv::IdQueue & job, Id ru
             // and negative for conflicts (conflicts with another package or dependency
             // conflicts = dependencies cannot be met).
             if (libsolv_solver.get_decisionlevel(p) > 0) {
-                q.push_back(p);
+                marked_for_install_providers_map.add_unsafe(p);
             } else {
-                Solvable * solvable = spool.id2solvable(p);
-                if (!spool.is_installed(solvable)) {
-                    available_unused_providers.push_back(solvable);
-                }
+                available_unused_providers_map.add_unsafe(p);
             }
         }
-        if (q.size() <= static_cast<int>(installonly_limit)) {
-            continue;
-        }
-        for (int k = 0; k < q.size(); ++k) {
-            Id id = q[k];
-            Solvable * s = spool.id2solvable(id);
-            if (spool->installed != s->repo) {
-                installing.push_back(id);
-                break;
-            }
+    }
+
+    libdnf5::solv::IdQueue marked_for_install_providers;
+    for (const auto & pkg_id : marked_for_install_providers_map) {
+        marked_for_install_providers.push_back(pkg_id);
+    }
+
+    std::vector<Solvable *> available_unused_providers;
+    for (const auto & pkg_id : available_unused_providers_map) {
+        Solvable * solvable = spool.id2solvable(pkg_id);
+        if (!spool.is_installed(solvable)) {
+            available_unused_providers.push_back(solvable);
         }
-        if (!installing.size()) {
+    }
+
+    const InstallonlyCmpData installonly_cmp_data{spool, running_kernel};
+    marked_for_install_providers.sort(&installonly_cmp, &installonly_cmp_data);
+    std::sort(available_unused_providers.begin(), available_unused_providers.end(), name_solvable_cmp_key);
+
+    libdnf5::solv::IdQueue same_names;
+    while (!marked_for_install_providers.empty()) {
+        Id name = same_name_subqueue(spool, &marked_for_install_providers.get_queue(), &same_names.get_queue());
+        if (same_names.size() <= static_cast<int>(installonly_limit)) {
             continue;
         }
 
-        const InstallonlyCmpData installonly_cmp_data{spool, running_kernel};
-        q.sort(&installonly_cmp, &installonly_cmp_data);
-        std::sort(available_unused_providers.begin(), available_unused_providers.end(), name_solvable_cmp_key);
+        // We want to avoid reinstalling packages marked for ERASE, therefore
+        // if some unused provider is also available we need to mark it ERASE as well.
+        auto low = std::lower_bound(
+            available_unused_providers.begin(), available_unused_providers.end(), name, name_solvable_id_name_cmp_key);
+        while (low != available_unused_providers.end() && (*low)->name == name) {
```

Comment:
I believe that there is a long standing problem, because we always calculate the order according to name, but may be we should add architecture. Think about `x86_64` and `i686`. Do you think whether it would be worth to add an architecture in account?

Reply:
This might sound strange, but handling of `available_unused_providers` by name is OK (or more secure), but handling of `marked_for_install_providers` by only name might be an issue.

Reply:
I think that would be another change in behavior so I would rather leave that for a different PR.

But I would be interested to know what kind of issues it might cause, perhaps a new github issue to track that would be nice. :+1: